File: parsers/python/fluxdsl/converters/clean_json.py
from ..json_ast import (
    Root, Pair, Include, Schema,
    Str, Num, Bool, Null, Obj, Lst,
    ObjPair, ObjInclude, ObjSchema, ObjEntry,
)
from ..flux_mapper import to_dto, from_dto
from ..parser import parse


def to_clean_value(val) -> object:
    """Convert a DTO value to a plain Python object (no type tags)."""
    if isinstance(val, Str):
        return val.value
    if isinstance(val, Num):
        return val.value
    if isinstance(val, Bool):
        return val.value
    if isinstance(val, Null):
        return None
    if isinstance(val, Obj):
        return _to_clean_obj(val)
    if isinstance(val, Lst):
        return [to_clean_value(e) for e in val.elements]
    raise TypeError(f"Unknown value type: {type(val)}")


def _to_clean_obj(obj: Obj) -> dict:
    result = {}
    for entry in obj.entries:
        if isinstance(entry, ObjPair):
            result[entry.key] = to_clean_value(entry.value)
        elif isinstance(entry, ObjInclude):
            _add_special(result, "$include", entry.path)
        elif isinstance(entry, ObjSchema):
            _add_special(result, "$schema", entry.path)
    return result


def _add_special(d: dict, key: str, value: str):
    if key in d:
        existing = d[key]
        if isinstance(existing, list):
            existing.append(value)
        else:
            d[key] = [existing, value]
    else:
        d[key] = value


def to_clean_root(root: Root) -> dict:
    """Convert a DTO Root to a plain dict (no type tags)."""
    result = {}
    for item in root.items:
        if isinstance(item, Pair):
            result[item.key] = to_clean_value(item.value)
        elif isinstance(item, Include):
            _add_special(result, "$include", item.path)
        elif isinstance(item, Schema):
            _add_special(result, "$schema", item.path)
    return result


def from_clean_value(value) -> object:
    """Convert a plain Python object back to a DTO value."""
    if value is None:
        return Null()
    if isinstance(value, bool):
        return Bool(value=value)
    if isinstance(value, (int, float)):
        return Num(value=value)
    if isinstance(value, str):
        return Str(value=value)
    if isinstance(value, list):
        return Lst(elements=[from_clean_value(v) for v in value])
    if isinstance(value, dict):
        return _from_clean_obj(value)
    raise TypeError(f"Unsupported value type: {type(value)}")


def _from_clean_obj(d: dict) -> Obj:
    entries: list[ObjEntry] = []
    for k, v in d.items():
        if k == "$include":
            _add_entries_special(entries, v, True)
        elif k == "$schema":
            _add_entries_special(entries, v, False)
        else:
            entries.append(ObjPair(key=k, value=from_clean_value(v)))
    return Obj(entries=entries)


def _add_entries_special(entries: list[ObjEntry], value, is_include: bool):
    cls = ObjInclude if is_include else ObjSchema
    if isinstance(value, list):
        for p in value:
            entries.append(cls(path=p))
    else:
        entries.append(cls(path=value))


def from_clean_root(d: dict) -> Root:
    """Convert a plain dict back to a DTO Root."""
    items = []
    for k, v in d.items():
        if k == "$include":
            if isinstance(v, list):
                for p in v:
                    items.append(Include(path=p))
            else:
                items.append(Include(path=v))
        elif k == "$schema":
            if isinstance(v, list):
                for p in v:
                    items.append(Schema(path=p))
            else:
                items.append(Schema(path=v))
        else:
            items.append(Pair(key=k, value=from_clean_value(v)))
    return Root(items=items)

File: parsers/java/src/main/java/fluxdsl/converter/CleanJson.java
package fluxdsl.converter;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.*;
import fluxdsl.json.JsonAst;
import java.util.*;

/** Converts between {@code JsonAst} DTOs and plain JSON/YAML trees (no type discriminators). */
public final class CleanJson {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private CleanJson() {}

    /** Converts a DTO root to a clean Jackson {@link ObjectNode}. */
    public static ObjectNode rootToClean(JsonAst.Root root) {
        var out = MAPPER.createObjectNode();
        for (var item : root.items()) {
            if (item instanceof JsonAst.Item.Pair p) {
                out.set(p.key(), valueToClean(p.value()));
            } else if (item instanceof JsonAst.Item.Include i) {
                addSpecial(out, "$include", i.path());
            } else if (item instanceof JsonAst.Item.Schema s) {
                addSpecial(out, "$schema", s.path());
            }
        }
        return out;
    }

    /** Converts a DTO value to a clean Jackson {@link JsonNode}. */
    public static JsonNode valueToClean(JsonAst.Value value) {
        if (value instanceof JsonAst.Value.Str s) return TextNode.valueOf(s.value());
        if (value instanceof JsonAst.Value.Num n) {
            var v = n.value();
            if (v == Math.floor(v) && !Double.isInfinite(v)) return new LongNode((long) v);
            return new DoubleNode(v);
        }
        if (value instanceof JsonAst.Value.Bool b) return BooleanNode.valueOf(b.value());
        if (value instanceof JsonAst.Value.Null) return NullNode.getInstance();
        if (value instanceof JsonAst.Value.Obj o) return objToClean(o);
        var l = (JsonAst.Value.Lst) value;
        var arr = MAPPER.createArrayNode();
        for (var e : l.elements()) arr.add(valueToClean(e));
        return arr;
    }

    /** Converts a DTO object value to a clean Jackson {@link ObjectNode}. */
    public static ObjectNode objToClean(JsonAst.Value.Obj obj) {
        var out = MAPPER.createObjectNode();
        for (var entry : obj.entries()) {
            if (entry instanceof JsonAst.ObjEntry.Pair p) {
                out.set(p.key(), valueToClean(p.value()));
            } else if (entry instanceof JsonAst.ObjEntry.Include i) {
                addSpecial(out, "$include", i.path());
            } else if (entry instanceof JsonAst.ObjEntry.Schema s) {
                addSpecial(out, "$schema", s.path());
            }
        }
        return out;
    }

    private static void addSpecial(ObjectNode node, String key, String value) {
        var existing = node.get(key);
        if (existing != null) {
            if (existing.isArray()) {
                ((ArrayNode) existing).add(value);
            } else {
                var arr = MAPPER.createArrayNode();
                arr.add(existing.asText());
                arr.add(value);
                node.set(key, arr);
            }
        } else {
            node.put(key, value);
        }
    }

    /** Converts a clean Jackson {@link ObjectNode} back to a DTO root. */
    public static JsonAst.Root cleanToRoot(ObjectNode node) {
        var items = new ArrayList<JsonAst.Item>();
        var it = node.fields();
        while (it.hasNext()) {
            var field = it.next();
            var key = field.getKey();
            var val = field.getValue();
            if (key.equals("$include")) {
                if (val.isArray()) {
                    for (var v : val) items.add(new JsonAst.Item.Include(v.asText()));
                } else {
                    items.add(new JsonAst.Item.Include(val.asText()));
                }
            } else if (key.equals("$schema")) {
                if (val.isArray()) {
                    for (var v : val) items.add(new JsonAst.Item.Schema(v.asText()));
                } else {
                    items.add(new JsonAst.Item.Schema(val.asText()));
                }
            } else {
                items.add(new JsonAst.Item.Pair(key, cleanToValue(val)));
            }
        }
        return new JsonAst.Root(items);
    }

    /** Converts a clean Jackson {@link JsonNode} back to a DTO value. */
    public static JsonAst.Value cleanToValue(JsonNode node) {
        if (node.isNull()) return new JsonAst.Value.Null();
        if (node.isBoolean()) return new JsonAst.Value.Bool(node.asBoolean());
        if (node.isNumber()) return new JsonAst.Value.Num(node.asDouble());
        if (node.isTextual()) return new JsonAst.Value.Str(node.asText());
        if (node.isArray()) {
            var elements = new ArrayList<JsonAst.Value>();
            for (var e : node) elements.add(cleanToValue(e));
            return new JsonAst.Value.Lst(elements);
        }
        if (node.isObject()) return cleanToObj((ObjectNode) node);
        throw new IllegalArgumentException("Unsupported node type: " + node.getNodeType());
    }

    /** Converts a clean Jackson {@link ObjectNode} back to a DTO object value. */
    public static JsonAst.Value.Obj cleanToObj(ObjectNode node) {
        var entries = new ArrayList<JsonAst.ObjEntry>();
        var it = node.fields();
        while (it.hasNext()) {
            var field = it.next();
            var key = field.getKey();
            var val = field.getValue();
            if (key.equals("$include")) {
                if (val.isArray()) {
                    for (var v : val) entries.add(new JsonAst.ObjEntry.Include(v.asText()));
                } else {
                    entries.add(new JsonAst.ObjEntry.Include(val.asText()));
                }
            } else if (key.equals("$schema")) {
                if (val.isArray()) {
                    for (var v : val) entries.add(new JsonAst.ObjEntry.Schema(v.asText()));
                } else {
                    entries.add(new JsonAst.ObjEntry.Schema(val.asText()));
                }
            } else {
                entries.add(new JsonAst.ObjEntry.Pair(key, cleanToValue(val)));
            }
        }
        return new JsonAst.Value.Obj(entries);
    }
}
